src/net/sourceforge/kolmafia/swingui/menu/LoadScriptMenuItem.java

```diff
@@ -13,25 +12,70 @@
  * In order to keep the user interface from freezing (or at least appearing to freeze), this
  * internal class is used to process the request for loading a script.
  */
-public class LoadScriptMenuItem extends DisplayFrameMenuItem {
+public class LoadScriptMenuItem extends ThreadedMenuItem {
 
   private final String scriptPath;
+  private boolean updateMRU = false;
 
   public LoadScriptMenuItem() {
     this("Load script...", null);
+    this.updateMRU = true;
+  }
+
+  public LoadScriptMenuItem(final File file) {
+    this(getRelativePath(file), getRelativePath(file));
+    this.updateMRU = true;
   }
 
   public LoadScriptMenuItem(final String scriptName, final String scriptPath) {
-    super(scriptName, "CommandDisplayFrame");
+    super(scriptName);
 
     this.scriptPath = scriptPath;
 
     this.addActionListener(new LoadScriptListener());
   }
 
-  @Override
-  public String toString() {
-    return this.scriptPath == null ? this.getText() : this.scriptPath;
+  public String getScriptPath() {
+    return this.scriptPath;
+  }
+
+  public static final String getRelativePath(File file) {
```

Comment:
For me, on Windows, the relative path of a relay script comes back as `/C:/KOLMAFIA_LOCATION/relay/SCRIPT_NAME.ash` instead of the `SCRIPT_NAME.ash` I think you intended.

Relay scripts don't get added any more by default (it seems) which I like, but I still see them for old scripts (and can make them appear if I call directly by absolute path).

Reply:
Hm, the logic was originally there to try to preserve as much of the old MRU menu items as possible without modifying their paths, but it doesn't make that much sense in hindsight.

I've simplified the relative path checks, so that ideally, your script would be remembered as `relay/SCRIPT_NAME.ash`. I don't know what relay scripts actually do, so if it's actually ideal that they are not included in MRU, it's a relatively simple fix to make them not be included there, and I can add that before I request another review.

Reply:
Two kinds of relay scripts:

- If you adventure in the relay browser, want to invoke xyx.php, and xyz.ash exists in the "relay" directory, you script is invoked. It has the opportunity to make the request and then mark up the response any way it wants. Additionally, for choice.php (and place.php and campground workshed items), you can have, for example, choice.1267.php to handle choice #1267 (genie wishing).
- Any script named relay_xxx.ash in the "relay" directory is added to the relay script dropdown on the right side of the topmenu. If you select such a script, it's a standalone app which can make any requests it wants and give you a fancy HTML front end. For example, relay_PVP_Bookkeeper.ash looks up your PVP fite history for the current season, looks at all the minis available, and gives you a table with your win/loss record for each of them, either offensively or defensively.

So, yeah - we don't want any of those appearing on any script history, since it makes no sense to invoke them by themselves.

That said, many such scripts import scripts that are in the scripts menu - and those  generally provide command line support to do this or that.

For example, Ezandora has Bastille Battalion relay support:

choice.1313.ash - choice.1319.ash are the various choices that support that IOTM.
Those import BastilleRelay.ash, which marks up the HTML (importing BastilleRelay.js to detect JS events).
It also calls upon scripts/bastille.ash, to do fancy calculations - which has a main function which parses a command line.

If you invoke bastille.ash directly, I expect it would go on a history list.
But when BastilleRelay imports bastille.ash, the latter's functions are built into the former's namespace, and calling those functions is not invoking the latter as a script.

Reply:
There was a lot of effort to make sure that everything listed in the MRU list was a script that could be launched from the Scripts menu.  If you could not launch it from there, then it would not be included and that includes relay scripts.  That said there are some bugs.  If you visit a page in the relay browser that has an associated relay override, the override script sometimes appears.  I have turned that into a feature since I find the MRU list serves as a "to do list" or a reminder that I have, or have not, done something.  But it does break the design contract.  Occasionally I find zlib on the list.  But many times I have typed something incorrectly in the gCLI and it has resolved as a request to invoke the script zlib.ash so the gCLI invoked it and hence it ended up on the list.  If one script invokes another using call, and perhaps something else I can't think of then it was included by design.  But if it was just included then it should not appear on the MRU list and if it does that is an unobserved or unreported bug :-)